pkg/term/proxy.go

```diff
@@ -31,48 +32,54 @@ func NewEscapeProxy(r io.Reader, escapeKeys []byte) io.Reader {
 	}
 }
 
-func (r *escapeProxy) Read(buf []byte) (int, error) {
-	nr, err := r.r.Read(buf)
+func (r *escapeProxy) Read(buf []byte) (n int, err error) {
+	if len(r.escapeKeys) > 0 && r.escapeKeyPos == len(r.escapeKeys) {
+		return 0, EscapeError{}
+	}
 
-	if len(r.escapeKeys) == 0 {
-		return nr, err
+	if len(r.buf) > 0 {
+		n = copy(buf, r.buf)
+		r.buf = r.buf[n:]
 	}
 
-	preserve := func() {
-		// this preserves the original key presses in the passed in buffer
-		nr += r.escapeKeyPos
-		preserve := make([]byte, 0, r.escapeKeyPos+len(buf))
-		preserve = append(preserve, r.escapeKeys[:r.escapeKeyPos]...)
-		preserve = append(preserve, buf...)
-		r.escapeKeyPos = 0
-		copy(buf[0:nr], preserve)
+	nr, err := r.r.Read(buf[n:])
+	n += nr
+	if len(r.escapeKeys) == 0 {
+		return n, err
 	}
 
-	if nr != 1 || err != nil {
-		if r.escapeKeyPos > 0 {
-			preserve()
+	for i := 0; i < n; i++ {
+		if buf[i] == r.escapeKeys[r.escapeKeyPos] {
+			r.escapeKeyPos++
+
+			if r.escapeKeyPos == len(r.escapeKeys) {
+				n = i + 1
+				err = EscapeError{}
+				break
```

Comment:
Wondering if it'd be clearer to return early here (possibly at the cost of some slight duplication)

Reply:
Here I think it could be replaced with:
``` go
n = i + 1 - r.escapeKeyPos
if n < 0 {
    n = 0
}
return n, EscapeError{}
```

Let me know if that looks better and I'll try it out.

Reply:
yes please! +1 to "return early"

Reply:
may be in a follow up PR is fine.